### PR TITLE
[Op][Training] Support Batch Norm Training Mode

### PR DESCRIPTION
**Pull request by SiriusNEO**
### Background

Operator `batch_norm` behaves differently in eval mode and train mode. In training, it will maintain two internal states `moving_mean` and `moving_var` while these updates don't appear in eval mode.

In [torch](https://pytorch.org/docs/stable/generated/torch.nn.functional.batch_norm.html), we have a flag `training: bool` to indicate this and a param `momentum` for updating. So it is necessary for us to introduce them to achieve training with bn.

What's more, consider we have a forward function with `batch_norm` operator in eval mode inside. If we have an AD pass, we may want to differentiate it and get the function which can be used to train. But the problem is that `batch_norm` is still in eval mode. So a pass is needed to convert all operators inside a function / mod from eval mode to training mode, by setting their flag `training` as `true`.

### Work

This PR adds `training: bool` and `momentum: float` arguments for batch norm in Relax and TOPI. And it contains a Pass named `ToTrainMode` which converts all operators to training mode.

### CODE REVIEW

File: python/tvm/relax/transform/legalize_ops/nn.py [hunk 1]
```diff
@@ -177,6 +177,10 @@ def _nn_batch_norm(bb: BlockBuilder, call: Call) -> Expr:
         epsilon=call.attrs.epsilon,
         center=call.attrs.center,
         scale=call.attrs.scale,
+        # By default relax batch_norm is training mode.
+        # To transform it to eval mode, use Norm SimplifyInference pass.
```
Comment (MasterJH5574) [inline python/tvm/relax/transform/legalize_ops/nn.py]: I suppose the name at this moment is SimplifyNormInference https://github.com/mlc-ai/relax/pull/120. Though it is not at the unity nor the mlc branch yet. Perhaps we can change the name now, and we'll have someone to commit that pass later.

File: python/tvm/topi/nn/batch_norm.py [hunk 2]
```diff
@@ -92,19 +102,47 @@ def batch_norm(
...
     shape[axis] = data.shape[axis]
 
-    moving_mean_rs = topi.reshape(moving_mean, shape)
-    moving_var_rs = topi.reshape(moving_var, shape)
-
-    out = (data - moving_mean_rs) / topi.math.sqrt(moving_var_rs + epsilon)
+    if training:
+        reduce_axes = list(range(len(data.shape)))
+        reduce_axes.remove(axis)
+        shape_prod = reduce(lambda x, y: x * y, [data.shape[ax] for ax in reduce_axes])
```
Comment (MasterJH5574) [inline python/tvm/topi/nn/batch_norm.py]: When the provided list is empty, the reduce operation will error. Would be great to confirm if that is possible, and add init value 1 if so.

**Review by MasterJH5574 — APPROVED** [review]

File: python/tvm/relax/transform/legalize_ops/nn.py (hunk 1 above)
Comment (SiriusNEO) [reply python/tvm/relax/transform/legalize_ops/nn.py]: Yes I originally intend to change the formal name when upstreaming the pass (Since I'm not sure whether it will still be this name), but it's also ok I modify here

File: python/tvm/topi/nn/batch_norm.py (hunk 2 above)
Comment (SiriusNEO) [reply python/tvm/topi/nn/batch_norm.py]: I think it's possible when data is 1-D. Let me add an init value.
